File: C-Plus-Plus/twin_prime.cpp
#include<iostream>
#include<vector>
using namespace std;

// helper function to check if a number if prime or not
bool isprime(int a){
  for(int i=2;i*i<=a;i++)
    if(a%i==0) return false;
  return true;
}

int main(){

  int a,b;
  cin>>a>>b;// input
  vector <int> prime;
  for(int i=a;i<b;i++){
    if(isprime(i))
      prime.push_back(i);
  }

  cout<<"Twin primes are: \n";
  for(int i=0;i<prime.size();i++){
    if(prime[i]+2==prime[i+1])
      cout<<prime[i]<<' '<<prime[i+1]<<endl;
  }

  return 0;
}
/*
twin prime are prime numbers which are two unit away from each other.

Implimentation:
  taking two input finding all prime in between them 
  then comparing adjacent numbers in vector if they differ by 2 or not  

Input:
  a and b specifing range

output:
  pair of twin prime

Sample Input-Output:
  Input1:
    1 10
  
  Output1:
    Twin primes are:
    3 5
    5 7

  Input2:
    1 100
  Output2:
  Twin primes are:
  3 5
  5 7
  11 13
  17 19
  29 31
  41 43
  59 61
  71 73

Time Complexity= O((a-b)*sqrt(b))
By prime number theory 
  number of prime number from 1 to n ~ n/log n
Space complexity = O(b/log(b)-a/log(a))
Contributed by : Hitansh K Doshi(github id: Hitansh159)
*/

File: Java/Algorithms/Twin_Prime.java
package Java.Algorithms;/*  Program to print all twin prime numbers present in a given range */

import java.util.*;

class Twin_Prime
{
    boolean isPrime(int n)
    {//Function to check if the integer is prime or not
        int c = 0;
        for(int i=1; i<=n; i++)
        {
            if (n%i == 0)
            c++;
        }
        if (c == 2)
            return true; //n is a prime number
        else
            return false; //n is a not prime number
    }

    void main()
    {
        Scanner sc = new Scanner(System.in);
        System.out.println("Please enter the start of the range:");
        int start = sc.nextInt();
        System.out.println("Please enter the end of the range:");
        int end = sc.nextInt();
        for (int i=start; i<=end; i++)
        {
            int a = i;
            int b = i+2;
            boolean r1 = isPrime(a);
            boolean r2 = isPrime(b);
            if ( r1 && r2)
            System.out.println(a + " " + b);
        }
    }
}


/*
Example:
Please enter the start of the range:
5
Please enter the end of the range:
29
5 7
11 13
17 19
29 31

Time Complexity: O(logn)
Space Complexity: O(1)
*/
